### ISSUE TITLE
License issues with the conda build

### ISSUE THREAD

**Issue by swails** [openmm/openmm#3171]
We recently encountered an issue with building and using OpenMM in a commercial setting.  While the license is commercial-friendly (LGPL for GPU parts and MIT for everything else, loosely), the incorporation of FFTW for the CPU-PME plugin renders, at a minimum, anything that dlopen's the CPU PME plugin GPL as well.

A fairly straightforward workaround is to link against Intel MKL instead of FFTW, as it has a commercial-friendly license and is a drop-in replacement wrt the FFT API that FFTW implements.  But currently as-released, the OpenMM conda package is forced into GPL (which most people may not realize).

**Comment by peastman** [comment]
Thanks for pointing this out.  FFTW also isn't strictly required.  If you don't have it, the only effect is that PME on the CPU platform will be slower.

**Comment by swails** [comment]
Right, but the default builds on conda-forge do have the plugin, and are automatically loaded into memory.

How do you feel about updating the conda builds to use the Intel MKL instead of FFTW?  I can share how we did the same thing building our conda package for it.  The 3 options I can see are:

1. The conda-forge builds become GPL
2. Drop the CPU-PME plugin for the conda build (to keep LGPL+MIT)
3. Build the conda package with the Intel MKL FFTW implementation instead of FFTW's.

**Comment by jchodera** [comment]
(3) sounds like an appealing option!

**Comment by peastman** [comment]
I think MKL is only for x86.  We also need ARM and PPC.

I suppose we could create special packages that don't list FFTW as a dependency, for people who need to avoid it?

**Comment by giadefa** [comment]
Just one question.

Is it allowed to link a GPL library to a non GPL code? Clib is GPL, isn't
it?

**Comment by peastman** [comment]
Any code you link to a GPL library has to be under a GPL-compatible license.  See https://www.gnu.org/licenses/license-list.en.html.  Technically speaking, at the time you link them (either in advance, or dynamically at runtime), they become joined into a single program, and because that program contains GPL code, the user is supposed to have the right to distribute the entire thing under the GPL.  A license is GPL-compatible if it allows you to use the code as part of a larger program that is GPL licensed.  MIT and LGPL are both GPL-compatible.

**Comment by giadefa** [comment]
But then any code linked against glibc would become GPL. Aren't the C
libraries GPL?

**Comment by giadefa** [comment]
Actually glibc is lGPL but the Linux kernel is still GPL.

**Comment by peastman** [comment]
https://opensource.stackexchange.com/questions/10223/should-i-publish-everything-running-on-linux-under-gpl

The Linux kernel adds a specific exception to say that if you invoke it through normal system calls, that doesn't count as linking and doesn't create a derived work.

**Comment by swails** [comment]
2 viable proposals:

1. Link MKL for supported architectures and fftw for unsupported ones, making sure to bundle the GPL license for FFTW-linked packages and MIT/LGPL for those linked to MKL
1. Bundle the PME plugin as a separate conda package that can be installed alongside OpenMM to enable accelerated PME. We can build separate MKL and FFTW variants available for the platforms they support (akin to how numpy supports MKL and BLAS flavors).

The first seems more pragmatic, the second seems more practical and transparent from a licensing perspective, and I can come up with arguments I find compelling for either option.

Advantages of the first approach:

* Transparent to the user - you don't "lose" optimized-PME with the new packages
* I suspect most commercial applications of OpenMM runs on amd64, so the importance of MIT/LGPL licensing is addressed automatically for the majority of places where it matters (pursuant to the validity of the initial assumption)

Disadvantages:

* You need to be careful that a commercial environment doesn't pull in FFTW (but arguably you should be careful about this regardless)

Advantages of the second approach:

* You can write an environment.yaml and rest assured you will get a commercial-friendly license no matter what
* You can strictly control whether you pull in MKL-bound CPU-PME or FFTW-bound CPU-PME for any (supported) hardware
* It is very explicit about what plugin is included
* You can add the plugin whenever you want and have it "automagically" work

------

Weighing all of this, my preference would fall on the first option.

**Comment by peastman** [comment]
> I suspect most commercial applications of OpenMM runs on amd64

The main exception to that is Macs, since they're now ARM.  If a company is selling a commercial product, though, they can control what platforms they sell it for.

The other option, of course, is that they don't rely on the standard conda package for OpenMM.  They bundle their own copy of it with their application.  That's what ChimeraX does.

**Comment by tristanic** [comment]
Would be a bit of extra work to adapt, but the C++ port of FFTPack at https://github.com/cctbx/cctbx_project/tree/master/scitbx/fftpack might be worth a look. Permissively licensed, well tested (used extensively in the Phenix suite), and about the same or slightly faster speed as FFTW3.

**Comment by jchodera** [comment]
This [discussion of FFT alternatives for scipy's backend](https://github.com/numpy/numpy/issues/11885) may also be illuminating.

**Comment by tristanic** [comment]
Good discussion [here](https://github.com/project-gemmi/benchmarking-fft), coming down strongly in favour of the [the cpp branch of PocketFFT](https://gitlab.mpcdf.mpg.de/mtr/pocketfft/-/tree/cpp). I gather Numpy and Scipy are both using it now, which means it's been getting some serious workout. Looking at using it to replace the MKL masquerading as FFTW2 in my own project.

**Comment by peastman** [comment]
Thanks for pointing that out.  We should look into it.  Since the whole thing is just a single header file and uses a BSD license, we could just bundle it in the `libraries` directory.  That would remove a lot of the complexity of linking to FFTW or MKL.

Based on the benchmarks at that page, the performance is competitive with FFTW for 3D FFTs as long as you compile with AVX.  (It's much slower for 1D and 2D, but we don't care about those.)  In #3544 we decided we want to continue supporting older processors for the moment, but I think it's ok if the PME plugin doesn't support them.  That just means it would fall back to the reference implementation of the reciprocal space calculation.  It's slower but it works, and hopefully no one expects the CPU platform to be fast on such old processors.

**Comment by peastman** [comment]
One other issue is that it uses the clang/gcc portable vector extension, which isn't supported by Visual Studio.  You can disable vectorization, but that would mean the CPU platform would be somewhat slower on Windows than other platforms.

**Comment by peastman** [comment]
I tried it out, and I'm getting performance significantly faster than FFTW.  Here are the times for the DHFR and ApoA1 benchmarks.  First on a 4-core x86 CPU.  I tried compiling with and without vectorization enabled, since we'll have to disable it on Visual Studio.

FFT|DHFR|ApoA1
---|---|---
FFTW|20.6|5.0
PocketFFT|28.4|6.6
PocketFFT, no vectorization|26.9|6.1

Disabling vectorization slows it down a little, but it's still much faster than FFTW.  I also tried on my laptop, which has a M1 Pro.  The difference is much smaller, but PocketFFT at least doesn't seem to be slower.

FFT|DHFR|ApoA1
---|---|---
FFTW|34.0|9.5
PocketFFT|35.4|9.4

I'll clean it up and create a PR.  This also can replace the FFT code we have in the reference platform.
